disnake/enums.py

```diff
@@ -72,326 +85,1370 @@
 )
 
 
-class _EnumValueBase(NamedTuple):
-    if TYPE_CHECKING:
-        _cls_name: ClassVar[str]
+_T = TypeVar("_T", bound="EnumMeta")
 
-    name: str
-    value: Any
 
-    def __repr__(self) -> str:
-        return f"<{self._cls_name}.{self.name}: {self.value!r}>"
+def _is_descriptor(obj):
+    return hasattr(obj, "__get__") or hasattr(obj, "__set__") or hasattr(obj, "__delete__")
 
-    def __str__(self) -> str:
-        return f"{self._cls_name}.{self.name}"
 
+class _EnumDict(Dict[str, Any]):
+    def __init__(self, base: Type[Any]):
+        # We explicitly take a base in __init__, unlike default Enums, to more easily
+        # enforce proper value typing. This should ensure better performance, as e.g.
+        # `Enum[int].one + 3 == 4` is more performant than `Enum[Any].one.value + 3 == 4`,
+        # as this saves a lookup. We do lose some functionality here, but it's probably
+        # worth the cost; especially considering all current enums have monotyped values.
+        super().__init__()
+        self.base = base
+        self.member_map: Dict[str, Any] = {}
+        self.value_map: Dict[Any, str] = {}
 
-@total_ordering
-class _EnumValueComparable(_EnumValueBase):
-    def __eq__(self, other: object) -> bool:
-        return isinstance(other, self.__class__) and self.value == other.value
+    def __setitem__(self, name: str, value: Any) -> None:
+        if name in {"mro", ""}:  # illegal names defined in original python enums
```

Comment:
should this also disallow `name` and `value`? Currently, a `name` property would prevent access to an enum member's underlying name.

Reply:
Yeah, I'd been wondering the same, but for the time being I kept it like this to mirror the vanilla enum implementation (at least, at the time of writing this pr). I'd personally agree with disallowing name and value, but on the other hand I wanted to keep the "api" as close to that of vanilla enums. Which would you prefer?

Reply:
Fair enough, yea.
The stdlib enums solve this through a custom descriptor iirc, but since we're mostly looking for speed and the descriptor isn't exactly fast, I'd say it's better to consider those two names illegal as well to avoid mishaps.

Reply:
Sounds good to me, will make this change when I get around to working on the pr.

Reply:
These were added in 662e0bdf1f0c47a3142d4214b1cbda16fb6e5b17, along with an extra check to accommodate for the new `.name` and `.value` descriptors.

Reply:
This was changed to direct access with a special `__setattr__` to protect `.name` and `.value` as part of 057f09b51e3ba2bc2fff37065f3072c0c940e378. The extra check can also be reverted again, which was done in 31757782843f0e01c9f3598b223faa8274f73264.